### ISSUE TITLE
Keyboard layout not detected.

### ISSUE THREAD

**Issue by bootkiller** [sddm/sddm#202]
SDDM isn't picking up the keyboard layout. 

Output of `/etc/vconsole.conf`:

```
KEYMAP=pt-latin9
FONT=Lat2Terminus16
```

Output of Systemd's  `localectl`

```
System Locale: LANG=en_GB.UTF-8
   VC Keymap: pt-latin9
   X11 Layout: pt
   X11 Model: dell101
   X11 Variant: latin9
```

**Comment by fgatbdx** [comment]
I have the same behavior with 2 new installation of Fedora 21, KDE spin. I did installation on 2 different PCs. Both installation were done in french.
 * On 1st PC the SDDM login window shows a US flag for the language, but as soon as I enter the password, the flag turns to French. Strange behavior but it works as mention by *tachylatus*.
 * The behavior is different on the second PC: at sddm login window, it is and remains US keyboard. The password is the same between PCs (same user too).

On both PCs 
```bash
LANG="fr_FR.UTF-8"
KEYMAP="fr-oss"
option "XkbLayout" "fr"
option "XkbVariant" "oss"
```
Boring bug if we want to use a safe password, with non std char.

**Comment by alexisju** [comment]
+1

**Comment by polomi** [comment]
Same thing on Arch Linux. Used localectl to configure the keymap for both the VC and X11. On the login screen, the language flag is set to US until I type the first character of the password and it switches to the right language by itself. Before typing the first character, while it still shows the US flag, the right language doesn't appear in the drop down menu, only US is present.

**Comment by akien-mga** [comment]
Exact same behaviour as what @polomi described above, on Mageia 5. FYI, here is the corresponding Mageia bug report: https://bugs.mageia.org/show_bug.cgi?id=15357

**Comment by miguelpenc** [comment]
Edit
```
/usr/share/sddm/scripts/Xsetup
```
and append
```
setxkbmap XX
```
where XX is the preferred layout.

**Comment by EvilMav** [comment]
@miguelpenc Thank you! That did the trick.

**Comment by davispuh** [comment]
This is Xcb bug (even with latest libxcb from git master branch), because in `XcbKeyboardBackend::initLayouts`
```c++
        cookie = xcb_xkb_get_names(m_conn,
                XCB_XKB_ID_USE_CORE_KBD,
                XCB_XKB_NAME_DETAIL_GROUP_NAMES | XCB_XKB_NAME_DETAIL_SYMBOLS);
        reply = xcb_xkb_get_names_reply(m_conn, cookie, nullptr);
```

it always returns "us" layout on startup, but as soon as any key is pressed then `XCB_XKB_NEW_KEYBOARD_NOTIFY` event will be issued and then it will return correct layout.

**Comment by Tratoschek** [comment]
A bonus would be, that the user somewhere can choose the layout at startup. 
Usecase: Consider a multiuser system, where some user prefer normal qwerty and another something like dvorak. Then the user(s), whose keyboard layout is not the default in sddm, has to learn to type his password in two layouts.

**Comment by majewsky** [comment]
@Tratoschek That's a separate issue and you should open a separate issue for that to ensure that it's not missed.

**Comment by Tratoschek** [comment]
done: https://github.com/sddm/sddm/issues/482

**Comment by Alko89** [comment]
I had the same issue and fixed it with localectl set-x11-keymap si asus_laptop. The keymap is now correct, but the dropdown still shows a US flag.

**Comment by akien-mga** [comment]
Hi there, any clue about this issue? It's mostly cosmetic but quite confusing for many users of non en_US locale to be greeted at every boot by an icon saying they are configured for en_US keyboard, while starting typing will switch to the actual configured locale.

**Comment by miguelpenc** [comment]
If it is just about cosmetics, then the easiest thing to do is either using a theme that doesn't show the language button, or modify one of your preference to the same purpose.

**Comment by akien-mga** [comment]
Dropping a feature completely for the sake of cosmetics doesn't feel right... And I'm looking here for a change that will be applied at the distro packaging level, if all distros start hacking the upstream theme to remove the keyboard choosing feature, it doesn't feel right either... :)

**Comment by doktor5000** [comment]
@davispuh By any chance, did you already report it upstream to libxcb?

**Comment by davispuh** [comment]
No, I haven't had time to investigate this more.

**Comment by doktor5000** [comment]
Taken the freedom to report it, let's see what comes out ...
https://bugs.freedesktop.org/show_bug.cgi?id=93701

**Comment by doktor5000** [comment]
@plfiorini Can you provide some help on how to run a binary and redirect it's output to some logfile
( code attached to https://bugs.freedesktop.org/show_bug.cgi?id=93701#c1 )
when sddm starts but **before** the first keyboard keypress? Would `/usr/share/sddm/scripts/Xsetup` work for that?

**Comment by miguelpenc** [comment]
@doktor5000 To my knowledge `/usr/share/sddm/scripts/Xsetup` works like that. to redirect output of a binary file use:
`~% ./name_of_the_program &> /path/to/logfile/logfiele.log`

**Comment by doktor5000** [comment]
@miguelpenc Well aware how to do shell redirection, I was just asking if this needs to be done from sddm-greeter, maybe my question wasn't phrased that well :)

Anyways, as sddm waits for `/usr/share/sddm/scripts/Xsetup` to exit before starting, didn't achieve exactly what I wanted, but it works. When `Xsetup` runs so before any key has been pressed, it returns `English (US)` and after a keypress or in the normal session it returns `German (eliminate dead keys)` as expected.

**Comment by rapgro** [comment]
Reproducible for me. Fedora 24 x86_64 with german locale.

**Comment by Compizfox** [comment]
I have the same problem I think.

Upon boot, SDDM is in QWERTY without a way to change that. It does not change to the correct layout (Colemak) when I start typing; I actually have to type my password in QWERTY to login.

    localectl
       System Locale: LANG=nl_NL.UTF-8
           VC Keymap: colemak
          X11 Layout: colemak

I also have `setxkbmap colemak` in `/usr/share/sddm/scripts/Xsetup` but that does not seem to have any effect.

**Comment by va1entin** [comment]
I can confirm that this is still an issue with SDDM 0.17
<pre>
$ localectl status
   System Locale: LANG=en_US.UTF-8
       VC Keymap: de
      X11 Layout: de
       X11 Model: pc104
     X11 Options: terminate:ctrl_alt_bksp
</pre>

SDDM still uses US layout and does not change to de when I type something

**Comment by Soukyuu** [comment]
So it seems that only very specific entries trigger the "switch on first keypress" behavior.
I've had a linux box configured as follows:
```
localectl status
   System Locale: LANG=en_US.UTF-8
       VC Keymap: de-latin1
      X11 Layout: de-latin1
```
and the keyboard in SDDM did NOT switch to de-latin1 when I typed in my password. Trying to debug the issue, I've set the keymap to `de-latin1-nodeadkeys` like this:
```
localectl set-keymap de-latin1-nodeadkeys
```
Note how I'm setting the vconsole keymap and not the X11 keymap! Now, the output is
```
localectl status
   System Locale: LANG=en_US.UTF-8
       VC Keymap: de-latin1-nodeadkeys
      X11 Layout: de
       X11 Model: pc105
     X11 Variant: nodeadkeys
     X11 Options: terminate:ctrl_alt_bksp
```
Note how the `VC Keymap` and `X11 Layout` differ, because localectl is automatically picking the closest match for X11? After a reboot SDDM now does switch to the correct keymap once I start typing my password.

So I'm thinking the problem might be a name mismatch between keymaps for X11 and VConsole at least for some of us...

**Comment by rtxrulez** [comment]
> Hi there, any clue about this issue? It's mostly cosmetic but quite confusing for many users of non en_US locale to be greeted at every boot by an icon saying they are configured for en_US keyboard, while starting typing will switch to the actual configured locale.

Yes. I have the same problem now. How did you decide it for yourself?

**Comment by rtxrulez** [comment]
> > Hi there, any clue about this issue? It's mostly cosmetic but quite confusing for many users of non en_US locale to be greeted at every boot by an icon saying they are configured for en_US keyboard, while starting typing will switch to the actual configured locale.
> 
> Yes. I have the same problem now. How did you decide it for yourself?

Writing to a file helped me `/usr/share/sddm/scripts/Xsetup` next line
```
setxkbmap us,ru

```

You can choose between layouts

**Comment by Compizfox** [comment]
I finally got it working by putting

    setxkbmap us -variant colemak

in `/usr/share/sddm/scripts/Xsetup`.

As I mentioned before, just

    setxkbmap colemak

doesn't have any effect.

**Comment by tim-hilt** [comment]
Doesen't work, when adding

```
localectl list-x11-keymap-layouts
```

to `/usr/share/sddm/scripts/Xsetup`. Still qwerty for me.

**Comment by flatmapthatshit** [comment]
Same issue here (openSUSE Leap 15.3).

Plus, apparently the English (US, intl, with dead keys) is broken.

Here's my `Xsetup`:

    #!/bin/sh
    # Xsetup - run as root before the login dialog appears
    setxkbmap -layout us,us,gb -variant ,intl,

And here's my `00-keyboard.conf`:

    Section "InputClass"  
        Identifier "system-keyboard"  
        MatchIsKeyboard "on"  
        Option "XkbLayout" "us,us,gb" 
        Option "XkbVariant" ",intl,"  
    EndSection

The KB layout switcher only appears after I type something in the password box. Plus, dead keys are not working properly. E.g., if I select English (US, intl, with dead keys), when I type ~ followed by a, the result is an a instead ã. If I type ' followed by a space it only displays the space.

Should I report his as a separate bug? (Assuming that this is an SDDM bug to begin with).

**Comment by davispuh** [comment]
> Same issue here (openSUSE Leap 15.3).
> 
> Should I report his as a separate bug? (Assuming thar this is an SDDM bug to begin with).

If you had read above you would see it's not SDDM bug, but it's a XKB bug that's still haven't fixed after so many years...

**Comment by flatmapthatshit** [comment]
> > Same issue here (openSUSE Leap 15.3).
> > Should I report his as a separate bug? (Assuming thar this is an SDDM bug to begin with).
> 
> If you had read above you would see it's not SDDM bug, but it's a XKB bug that's still haven't fixed after so many years...

The part about late layout initialisation sure, but I'm not certain that this is the root cause for dead keys not to be working (that's what I'm asking if I should report as a separate issue).

Can anyone try adding this to Xsetup

    setxkbmap -layout us -variant intl

And check if they can type  combinations such as ã (~ + a) and " (" + space)?

To me `setxkbmap -layout us -variant intl` is certainly doing something... As in, I get different behaviours with and without the variant. Nevertheless the intl variant is not working properly in SDDM (it does works in Plasma).

**Comment by davispuh** [comment]
Ohhh indeed, I missed that part, that's a completely different issue so definitely submit a new issue.

**Comment by maxnikulin** [comment]
On Debian and Ubuntu `/etc/X11/xorg.conf.d/00-keyboard.conf` may be missed, keyboard configuration is stored in `/etc/default/keyboard` and `localectl` is able to report it. So I would not rely on X config files.

I faced an obscure issue that X session *may be* started with US only layout (accordingly to `setxkb -print`) ignoring layout switch, compose options, etc. when opened by SDDM. My expectation is that global settings should be used unless the user explicitly chooses another configuration somehow.

**Comment by tuxslack** [comment]
# Dica testada no Void Linux em 2024


Para gera o arquivo de configuração do sddm no  Void Linux

Root:

 sddm --example-config > /etc/sddm.conf


Solução para layout do teclado em BR:

$ cat /usr/share/sddm/scripts/Xsetup | grep setxkbmap
setxkbmap br


Root:

 gedit /etc/sddm.conf


Verificar o tema:

$ sddm-greeter --test-mode --theme /usr/share/sddm/themes/maldives


Reiniciar o sistema

Root:

 reboot

**Comment by cl0ne** [comment]
@tuxslack what's that all about?

**Comment by mdluhosch** [comment]
Hi,
I found the issue in the XServer but unfortunately I don't think it will be fixed. Have a look at: 
https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/1525
Maybe you should think about implementing Peters suggestion.

**Comment by tuxslack** [comment]
> Hi, I found the issue in the XServer but unfortunately I don't think it will be fixed. Have a look at: https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/1525 Maybe you should think about implementing Peters suggestion.

In Void Linux there is no **localectl** in the repositories

**Comment by tuxslack** [comment]
> @tuxslack what's that all about?

Palliative method:

To show the keyboard layout in your language in the sddm login manager. It worked for BR.

After changing, test or start the system

sddm-greeter --test-mode --theme /usr/share/sddm/themes/maldives

**Comment by tuxslack** [comment]
> Hi, I found the issue in the XServer but unfortunately I don't think it will be fixed. Have a look at: https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/1525 Maybe you should think about implementing Peters suggestion.



> @tuxslack what's that all about?

Palliative method:

To show the keyboard layout in your language in the sddm login manager. It worked for BR.

After changing, test or start the system

sddm-greeter --test-mode --theme /usr/share/sddm/themes/maldives

**Comment by Vogtinator** [comment]
> Maybe you should think about implementing Peters suggestion.

What is his suggestion?

We could use libxkbregistry or parse python-langtable info to list possible keyboard layouts. Applying it on X11 is a bit annoying but possible, for Wayland there is still no API (weston doesn't allow changing the layout during runtime, kwin has no runtime API but only listens for config file changes)

**Comment by mdluhosch** [comment]
@tuxslack Sorry it was a bit misleading I was talking about https://github.com/sddm/sddm/issues/202#issuecomment-133628462


@Vogtinator 
> 
> What is his suggestion?
> 
> We could use libxkbregistry or parse python-langtable info to list possible keyboard layouts. Applying it on X11 is a bit annoying but possible, for Wayland there is still no API (weston doesn't allow changing the layout during runtime, kwin has no runtime API but only listens for config file changes)

He wrote:

> For ssdm it'd be better to have its own config and apply the keyboard layout on startup so it doesn't have to guess what layout the server initializes with (and which device). It feels this approach here is just a hack that's good enough for some but doesn't actually fix the real issue.

and

> Anyway, the summary is: we've had dynamic input devices for almost 20 years, we've had the current issue for about 16 years now and it cannot be correctly worked around in the X server. The _only_ way to fix it is to have the desktop environment handle the keymaps (which they do with Wayland as well btw).

So basically he suggested to use an own config system for the layout(s), similar to how it is done in kde or gnome
